WHF-213: Expose MembersOnlyEventAccess.get API to check for groups-only events

ahed-compucorp opened the pull request:
## Overview
This PR is to expose `MembersOnlyEventAccess.get` API to gets event access details for a given event ids, this is useful to check for groups-only events.

## Technical Details
The API returns event access details as follows: 
```php
[
  'event_id' => 1,
  'is_groups_only_event' => TRUE,
  'allowed_groups' => [1, 2],
  'does_user_belongs_to_any_allowed_group' => TRUE,
  'allowed_groups_which_user_belongs_to' => [1],
]

jamienovick commented:
@ahed-compucorp @ana-compucorp 
Hey guys, I don't think we should be listing events for groups only in the SSP if users cannot book on them. I'd guess they should only show for users who can book on them? Perhaps we can get some better visibility of the use cases from client? What types of events are these and should people in SSP who cannot book on them know they exist?
Otherwise we should introduce a setting so we can be flexible.

@madeleinevironda FYI - perhaps you have some insight? Did client sign this point off?
Thanks
J

madeleinevironda commented:
@jamienovick - this is what the client requested, they want SSP users to see all events that are taking place even if they can't sign up for them as a form of marketing.

ivan-compucorp reviewed (commented):
@ahed-compucorp you've implemented a very interesting idea of using drupal hooks inside civicrm, in general I like the approach, maybe civicrm would have something similar in future.
At the same time I was thinking if it's correct to do some processing for drupal views like this. In theory `uk.co.compucorp.membersonlyevent` is a civicrm extension that can be installed on any civicrm site, hence it should not be aware of views (blocks, nodes, etc) or other entities that drupal has (in ideal world civicrm should run on other cms as well). So using some particular view names like `ssp_events` and especially view field names like `nothing` is not appropriate here as this code does not make any sence for all other sites (except ssp).
I can see 2 ways of solving a problem that we have:
1. Keep some code in civicrm (check if event is group only) and alter the field output in drupal (update `ssp_core_events_views_field_alter_events_listing()` respectively).
1. Do all processing on drupal side (update `ssp_core_events_views_field_alter_events_listing()` respectively). If we want to save some performance we can use some other hook instead of `hook_preprocess_views_view_field` to run 1-2 queries to check all events in the list at once.

I assume option 1 is (almost) impossible, hence suggest to go ahead with option 2. What do you think?

cc @ana-compucorp

ahed-compucorp commented:
> I was thinking if it's correct to do some processing for drupal views like this. 

I'll reply with two points to make it clear.
1- You are making sense about updating the SSP views from SSP code instead of the extension's code.
I did it this way because of the specification.
cc @madeleinevironda 

2- Should we use particular view/display names?
You're 100% right. I've updated the code to check for the module and table names instead - see [8f1ce8](https://github.com/compucorp/uk.co.compucorp.membersonlyevent/commit/8f1ce824f06272911dfc77482ea4bbf530f4013e#diff-6393b3e2eb1a53220b37d60ebec10313dde2e6da6542f7f7b65b411494291c5bR30)

> we can use some other hook instead of hook_preprocess_views_view_field to run 1-2 queries.
Yes, I've updated the code to use the `hook_views_post_execute` to check all events in the list at once - see [dd0bb1](https://github.com/compucorp/uk.co.compucorp.membersonlyevent/commit/dd0bb11c150b9b7b5a526e9e532185d597f61289).

ahed-compucorp commented:
@erawat
Unfortunately , It's Drupal only.

jamienovick commented:
@ahed-compucorp - @ana-compucorp asked me to take a quick look.

@erawat might have a different view on this and I may be missing something but I think conceptually we should make the CiviCRM extensions independent of the SSP. The CiviCRM extensions could be installed on any CMS as @erawat mentioned (i.e. Wordpress, D8/9, Backdrop) but also could be installed without the SSP. As such there shouldn't be anything within the extension to create a dependancy on SSP.

It would however be totally appropriate for SSP to depend on CiviCRM. In some ways the SSP sites "on top" of CiviCRM and hence can have various dependancies there. As such I think this functionality should actually be part of SSP.

As such that would probably mean that SSP would require this CiviCRM extension being installed.

If we want to make things more complex perhaps we can check in the code IF this CiviCRM extension is enabled and only if it is enabled to then do the appropriate additional processing.

erawat commented:
> @ahed-compucorp - @ana-compucorp asked me to take a quick look.
> 
> @erawat might have a different view on this and I may be missing something but I think conceptually we should make the CiviCRM extensions independent of the SSP. The CiviCRM extensions could be installed on any CMS as @erawat mentioned (i.e. Wordpress, D8/9, Backdrop) but also could be installed without the SSP. As such there shouldn't be anything within the extension to create a dependancy on SSP.
> 
> It would however be totally appropriate for SSP to depend on CiviCRM. In some ways the SSP sites "on top" of CiviCRM and hence can have various dependancies there. As such I think this functionality should actually be part of SSP.
> 
> As such that would probably mean that SSP would require this CiviCRM extension being installed.
> 
> If we want to make things more complex perhaps we can check in the code IF this CiviCRM extension is enabled and only if it is enabled to then do the appropriate additional processing.

I totally agree with Jamie, if we want to hide Book Now button from SSP, perhaps, the logic should be in one of the SSP module (e.g. SSP Membes only event) which check the condition of each event.

ahed-compucorp commented:
@erawat 
Moved the logic related to hiding the "Book Now" button to SSP - see this https://github.com/compucorp/ssp_core/pull/214 PR.
Updated this PR to only check if the user can register for the event and to expose a function to be used outside of CiviCRM world.

erawat commented:
@ahed-compucorp  Should we consider providing an API from Members Only Event extension that provides ability to check if an event is a members only event instead of exposing the function so other modules / plugin won't reline on the function result but relies on standard CiviCRM API's response?
